staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shufflesharding
+
+import (
+	"errors"
+	"math"
+)
+
+const maxHashBits = 64
+
+// ValidateParameters can validate parameters for shuffle sharding,
+// including queueSize and handSize
+// Algorithm: maxHashValue >= queueSize!/(queueSize-handSize)!
+func ValidateParameters(queueSize, handSize uint32) bool {
+	if handSize == 0 || queueSize == 0 || handSize > queueSize {
+		return false
+	}
+
+	maxHashValue := uint64(1<<maxHashBits - 1)
+	for i := uint32(0); i < handSize; i++ {
+		maxHashValue /= uint64(queueSize - i)
+		if maxHashValue == 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+// FastValidateParameters can validate parameters for shuffle sharding
+// in a fast but rough way, including queueSize and handSize
+// Algorithm: maxHashValue >= queueSize^handSize
+func FastValidateParameters(queueSize, handSize uint32) bool {
+	if handSize == 0 || queueSize == 0 || handSize > queueSize {
+		return false
+	}
+
+	return math.Log2(float64(queueSize))*float64(handSize) <= maxHashBits
+}
+
+// ShuffleSharding can shuffle a hash value to handSize-quantity and non-redundant
+// indices of queue
+// Eg. From queueSize=128, handSize=8, we can get an index array [12 14 73 18 119 51 117 26]
+// Algorithm: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#queue-assignment-proof-of-concept
+func ShuffleSharding(hashValue uint64, queueSize, handSize uint32) ([]uint32, error) {
```

Comment:
If we fold the choosing into this function then the expense of allocating slices can be avoided.

Reply:
I agree it makes sense to simply return the queueIndex with the least packets here vs giving a hand and having the shortest queue selected after the fact, esp. given that would save slice allocations.

Reply:
```suggestion
func ShuffleSharding(hashValue uint64, queueSize, handSize uint32, f func(hand int)) error {
```

how about this? we put the choosing logic into `func(hand int)`

Reply:
ideally this package should a general utility for shuffle-sharding algorithm, its input doesnt have to be `queueSize` and `handSize`, how about we rename the params?

Reply:
(1) It was never "queue size", it is "number of queues".  Replacing that with "number of cards" would be fine.

(2) We definitely want a func parameter, so this can be tested in isolation.  For the proposed func, the parameter is a single card rather than a whole hand.

(3) The original ShuffleDealAndChoose function needs to be modified so that `lengthOfQueue` is a parameter, and could be factored into (a) a ShuffleAndDeal function that takes an `examineCard func(card int)` parameter and (b) a ShuffleDealAndChoose func that takes a `lengthOfQueue func(int) int` parameter and invokes ShuffleAndDeal passes `examineCard: func(card int) {qlen := lengthOfQueue(card); if qlen < bestQueueLen {...}}`.